### PR TITLE
feat: add isAssociated system contract

### LINKED ISSUE #8082: Add HIP-719 isAssociated() function call support
### Problem

The `isAssociated()` call as described in HIP-719 is not currently supported. 

### Solution

Add support for the `isAssociated()` call as described in HIP-719. This involves performing the following tasks

- [ ] Move notion design page https://www.notion.so/swirldslabs/HIP-719-isAssociated-11d7218461b64cefbd3371fd9afaa2bf into a design document
- [ ] Update the answerGiven method in RedirectViewExecutor to include handling for the isAssociated function selector (0x4d8fdd6d).
- [ ] Create a new query token precompile using an existing precompile as a model (e.g. NamePrecompile, BalanceOfPrecompile, or SymbolPrecompile) to handle calls to isAssociated.
- [ ] write appropriate tests

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by mustafauzunn**
**Description**:

Add support for the isAssociated() call as described in HIP-719.

**Related issue(s)**:

Fixes #8082 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    public static final Function IS_ASSOCIATED = new Function("isAssociated()", ReturnTypes.BOOL);
+
+    @Inject
+    public IsAssociatedTranslator() {
+        // Dagger2
+    }
+
+    @Override
+    public boolean matches(@NonNull HtsCallAttempt attempt) {
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java:42]: probably want to mark this as final here and below

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedCall.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import com.hedera.node.app.service.contract.impl.exec.gas.SystemContractGasCalculator;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.AbstractRevertibleTokenViewCall;
+import com.hedera.node.app.service.contract.impl.hevm.HederaWorldUpdater.Enhancement;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+public class IsAssociatedCall extends AbstractRevertibleTokenViewCall {
+    private final AccountID sender;
+
+    public IsAssociatedCall(
+            SystemContractGasCalculator systemContractGasCalculator,
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedCall.java:36]: nit: missing @NonNull annotations and finals

**Comment by lukelee-sl** [comment]
A couple of nits but looks good once those bdd tests can be added.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/IsAssociatedSystemContractTest.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+                            .initialSupply(0)
+                            .supplyKey("multikey")
+                            .exposingCreatedIdTo(nonFungibleTokenNum::set));
+        }
+
+        @HapiTest
+        @DisplayName("when token is not associated with a contract")
+        public Stream<DynamicTest> checkTokenIsNotAssociatedWithContract() {
+            return hapiTest(
+                    sourcing(() -> contractCall(
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/IsAssociatedSystemContractTest.java:131]: Is the lambda necessary?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/IsAssociatedSystemContractTest.java [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+                                                    getABIFor(
+                                                            Utils.FunctionType.FUNCTION, "isAssociated", "HRCContract"),
+                                                    isLiteralResult(new Object[] {false})))));
+        }
+
+        @HapiTest
+        @DisplayName("when nft is not associated with a contract")
+        public Stream<DynamicTest> checkNftIsNotAssociatedWithContract() {
+            return hapiTest(
+                    sourcing(() -> contractCall(
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/IsAssociatedSystemContractTest.java:153]: Many of these tests have basically the same structure.  Is it possible to create a static helper which is largely the same and just call that with a few parameters?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    public static final Function IS_ASSOCIATED = new Function("isAssociated()", ReturnTypes.BOOL);
+
+    @Inject
+    public IsAssociatedTranslator() {
+        // Dagger2
+    }
+
+    @Override
+    public final boolean matches(@NonNull HtsCallAttempt attempt) {
```
Comment (MrValioBg) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java:42]: nit:
```suggestion
    public final boolean matches(@NonNull final HtsCallAttempt attempt) {
```

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        return (attempt.isTokenRedirect() && matchesHrcSelector(attempt.selector()));
+    }
+
+    @Override
+    public final Call callFrom(@NonNull HtsCallAttempt attempt) {
```
Comment (MrValioBg) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/isassociated/IsAssociatedTranslator.java:47]: nit:
```suggestion
    public final Call callFrom(@NonNull final HtsCallAttempt attempt) {
```

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/isassociated/IsAssociatedTranslatorTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.isassociated.IsAssociatedTranslator;
+import com.hedera.node.app.service.contract.impl.hevm.HederaWorldUpdater.Enhancement;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class IsAssociatedTranslatorTest {
```
Comment (MrValioBg) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/isassociated/IsAssociatedTranslatorTest.java:38]: nit: The public modifier for the class and the methods, could be removed.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/isassociated/IsAssociatedCallTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.isassociated.IsAssociatedCall;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.isassociated.IsAssociatedTranslator;
+import com.hedera.node.app.service.contract.impl.test.exec.systemcontracts.common.CallTestBase;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.junit.jupiter.api.Test;
+
+public class IsAssociatedCallTest extends CallTestBase {
```
Comment (MrValioBg) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/isassociated/IsAssociatedCallTest.java:30]: nit:
```suggestion
class IsAssociatedCallTest extends CallTestBase {
```

**Review by MrValioBg — APPROVED** [review]
TYVM! I just left some nit comments that you could consider.
LGTM!

**Review by stoyanov-st — APPROVED** [review]
LGTM 💯!  +1 to @MrValioBg's comments

**Review by MrValioBg — APPROVED** [review]
LGTM!

**Review by stoyanov-st — APPROVED** [review]

**Pull request merged**